### ISSUE TITLE
Ability to upgrade all existing plugins by install-plugins.sh

### ISSUE THREAD

**Issue by szpak** [jenkinsci/docker#741]
When a pre-configured base Jenkins image is used (e.g. from OpenShift), it contains some plugins in particular versions. To be able to upgrade them to the latest possible version on a Docker image build time, it is required to generate a list of all installed plugins (base plugins + their dependencies) and add it to a list of plugins past to install-plugins.sh (paying attention to do not interfere with the plugin which intentionally should stay with particular versions) during an image building phase. It's tedious.

It would be convenient to have a switch in `install-plugins.sh` (`-u`?) to:
 - automatically update all existing plugins (when executed `install-plugins.sh -u`)
 - install the new plugins requested plugins and upgrade all existing plugins (paying attention to respect plugin versions on a passed list) (and executed `install-plugins.sh -u < list-of-plugins-to-install.txt`).

**Comment by oleg-nenashev** [comment]
https://github.com/jenkinsci/incrementals-tools#updating-versions-for-jenkins-docker-images is a starting point for that, but it handles only incrementals versions nowadays. trivial to extend tho

**Comment by geerlingguy** [comment]
I'm running into this issue now with the 2.166 version of the image; it has Script Security 1.53 installed, but Matrix Project Plugin version 1.14 requires 1.54 or later for security reasons.

Therefore the Matrix plugin gets disabled and then I get some additional security warnings. This is with a very vanilla setup using the following `plugins.txt`:

```
git
ansicolor
role-strategy
build-timestamp
parameterized-scheduler
kubernetes
```

So it would be _really_ nice if it were easy to update all previously-installed plugins as part of the image build process.

**Comment by geerlingguy** [comment]
Hmm, possibly related: https://github.com/jenkinsci/docker/issues/790

**Comment by geerlingguy** [comment]
So in my case, it looks like I was getting confused by Jenkins' behavior with regard to plugins in Docker images. I build a fresh Jenkins image for my infra on a regular basis, and I _thought_ that the plugins that were installed on that version of the image were current, and used in the running image.

But apparently Jenkins will use the existing plugins if they are being loaded from a pre-existing JENKINS_HOME and not update from the new versions in this container build.

So I have to go into my containerized Jenkins environments and manually update plugins using the UI.

**Comment by lifeofguenter** [comment]
Exactly my issue @geerlingguy - actually it is now causing plugins to lose their settings :sob:

**Comment by torstenwalter** [comment]
If #577 would be merged that you could simply set `PLUGINS_FORCE_UPGRADE` to `true`

**Comment by carlossg** [comment]
Fixed with #577

**Comment by mangitkaur** [comment]
Hi @geerlingguy : could you please explain the workaround here, as in my case also jenkins image has the upgraded plugin , but its nit been picked up by the container?
